[TEAM-24] [땃쥐, Lee] BE 세번째 PR입니다.

ttasjwi opened the pull request:
## 안녕하세요!
[TEAM-24] [땃쥐, Lee] 백엔드 3번째 PR입니다.

안녕하세요 브라이언! 24조 백엔드 땃쥐, Lee입니다.
3번째 PR을 보내드립니다.

## 배포 URL
http://15.165.236.13

## API 명세
https://github.com/sungju-kim/sidedish/wiki/%EB%B0%B1%EC%97%94%EB%93%9C-API-%EB%AA%85%EC%84%B8

## 작업 내역
- [x] `Item`, `Order`, `Delivery의` 저장계층(Repository) 구현
    - `Item`의 경우 요구사항에 상품 추가 로직이 없어서 `save()` 메서드는 따로 구현하지 않고, 하드코딩한 더미 데이터를 활용해 클라이언트와 연결하도록 구현했습니다.
    - `OrderRepository`의 경우 자바 코드상에서 필드로 들고 있는 객체들도 Select 해 오면서 생성해 주어야 해서 여러 빌더 메서드들이 붙게 되었는데요. 저희가 Spring Data JDBC를 사용하지 못하고 일반 JDBCTemplate을 사용하면서 생긴 불편함인 것 같습니다. JPA를 조금 더 중점적으로 학습하며 이후 미션에서는 이런 부분들을 개선해나가려고 합니다.
- [x] `ItemService` 클래스 구현
    - `itemId`를 인자로 받아 해당 `id`값에 해당하는 `ItemDetailResponse` DTO를 리턴하는 `findOne()` 메서드와, 인자로 들어온 섹션에 해당하는 `ItemListResponse` DTO를 리턴하는 `findBySection(...)` 메서드를 구현했습니다.
    
---

## 현재 코드에서 부족하다고 느끼는 것들
- 시간부족으로 도메인의 일부 로직들에 대한 테스트코드를 작성하지 못 했습니다.
- 시간부족으로 당장 기능에서 필요한 로직들에 해당하는 메서드들만 구현했습니다. 제대로 된 테스트를 하려면 모든 엔티티들에 대한 Save, Find 메서드를 구현해야했습니다. 실제로 테스트코드에서는 트랜잭션에서 테스트상품을 추가하고 그것을 기반으로 테스트를 해야했는데, data.sql에서 등록한 요소들의 존재를 전제로하여 테스트가 진행되고 있습니다.
- 기능구현에서 충분한 예외처리가 수행되지 못 하고 있다. 예외를 던지기만 하지, 예외를 어떻게 처리할지는 구현하지 못 하고 있다. 인터셉터 등 Spring에서 지원하는 예외처리 기능을 충분히 학습해야겠습니다.
- Spring-Data-Jdbc을 사용했으면 편리하게 처리할 수 있었을 법한 로직들도 있었으나, 당장 이에 대해 익숙하지 못 해 그냥 쿼리를 손수 짜서 날렸습니다...
- 상품과 상품이미지 사이에는 1:多 관계가 있는데 이를 SELECT 해올 때 rowMapper 메서드에서 내부적으로 이미지 목록을 가져오는 쿼리를 날려서 가져와 rowMapping 해주는 식으로 처리했습니다. 이를 어떻게 처리했으면 좋을지는 아직 고민 중입니다.

---

street62 committed:
[#38] Refactor: DDL 구조 개선

기존 테이블 설계에서 미흡했던 테이블 간의 외래키 참조관계를 정리하고, 추가적으로 필요한 테이블 정의 추가.

street62 committed:
[#39] Fix: Delivery 접근 제어자 누락 수정

deliveryType과 address에 누락된 접근 제어자(private) 추가.

street62 committed:
Style: DiscountPolicy initDiscountPolicyId() 메서드명 수정

street62 committed:
[#26] Feat: 상품 섹션에 해당하는 ItemSection 클래스 정의

street62 committed:
[#27] Feat: ItemImage, ItemImageType 정의

상품 이미지에 해당하는 ItemImage 클래스와, 이미지의 종류에 해당하는 ItemImageType 클래스 정의.

street62 committed:
[#28] Feat: 상품에 해당하는 Item 클래스 정의

street62 committed:
[#29] Feat: 주문에 해당하는 Order 클래스 정의

street62 committed:
[#41] Feat: ItemRepository 구현

상품 데이터의 저장 계층인 ItemRepository 구현.

street62 committed:
[#41] Test: ItemRepository 테스트 코드 작성

ItemRepository 클래스의 findById(), findBySectionName() 테스트 코드 작성.

street62 committed:
[#43] Feat: DeliveryRepository 구현

배송 저장 계층인 DeliveryRepository 구현.
배송을 DB에 저장하는 save(), id값에 해당하는 배송을 반환하는 findById() 메서드 작성.

street62 committed:
[#43] Test: DeliveryRepository 테스트 작성

save(), findById() 메서드 단위 테스트 작성.

ttasjwi committed:
[#42] Feat: 주문 저장 계층 OrderRepository 구현

주문 저장 계층 OrderRepository를 구현했다.
- 주문 등록(save)
- 주문 조회(findById)

ttasjwi committed:
[#42] Test: OrderRepository 테스트코드 구현

- saveAndFindTest : 주문을 저장하고 같은 주문 Id로 조회했을 때, 같은 주문이 찾아져야한다.

ttasjwi committed:
[#45] Feat: Member 클래스에 마일리지 획득 메서드 추가

- Member 클래스에서 매개변수로 들어온 수 만큼 해당 인스턴스의 mileage 값을 증가시키는 메서드 작성.(earnMileage(...))

ttasjwi committed:
[#45] Test: Member - earnMileage 메서드에 대한 테스트 메서드 추가

- earnMileage 메서드를 호출 후 member의 mileage값이 정상적으로 변경됐는 지 여부를 테스트하도록 함.

ttasjwi committed:
[#44] Feat: MemberRepository - 회원 정보 갱신 메서드 추가

- MemberRepository에 회원 정보 갱신기능을 담당하는 update 메서드 추가

ttasjwi committed:
[#44] Test: MemberRepository의 update 메서드 테스트 메서드 추가

- 회원정보를 변경시킨 후 update() 메서드를 통해 회원 정보를 갱신시킨 뒤, findById() 메서드로 조회하면 갱신된 회원이 반환되는지 테스트

ttasjwi committed:
[#47] Feat: Item 도메인에 새벽배송 지원 여부 필드 추가

Item 도메인에 새벽배송 지원 여부를 의미하는 supportDawnDelivert 필드를 추가

- Item 객체 수정 및 Repository(Item, Order) 코드 수정
- schema.sql 테이블 생성 쿼리 수정
- 샘플 더미 데이터 생성 쿼리인 data.sql 수정

street62 committed:
[#48] Feat: Item 클래스에 재고를 감소시키는 로직 추가

인자로 들어온 int reduceQuantity만큼 재고 수량을 감소시키는 reduceStock() 메서드와, 해당 메서드 내부에서 인자가 현재 재고 수량 이하인지 검증하는 validateReduceQuantity() 메서드 추가.

street62 committed:
[#48] Test: Item 클래스 재고 수량 감소 메서드 테스트 추가

reduceStock() 메서드를 호출해 재고 수량이 올바르게 감소하는지, 인자가 재고 수량보다 클 때 예외가 발생하는지를 검증하는 테스트 작성.

street62 committed:
[#49] Feat: Item의 재고 수량을 업데이트하는 updateStock 메서드 추가

UPDATE 쿼리와 jdbcTemplate.update() 메서드를 통해 아이템의 재고 수량을 업데이트하는 메서드 추가.

street62 committed:
[#49] Test: ItemRepository의 재고수량 업데이트 메서드 테스트 작성

updateStock() 메서드 호출 시 올바르게 재고 수량이 업데이트되는지 테스트 작성.

street62 committed:
[#52] Feat: Delivery 클래스의 배송료 관련 상수 추가

Delivery 무료요금, 기본요금, 무료요금에 대한 최소구매액 상수 추가
기본요금 2500원, 40000원 이상 구매 시 무료배송.

street62 committed:
[#51] Fix: DiscountPolicy의 할인금액 계산로직 반환타입 int로 변경

street62 committed:
[#50] Feat: Item에 현재 상품할인가격, 적립금 계산 로직 추가

- 상품 할인가격 : 할인정책, 상품 원가를 기반으로 할인가격 계산
- 적립금 계산 : 상품 할인가격의 1%를 적립금으로 함

street62 committed:
[#31] Feat: 상품 id로 상품 상세정보를 조회하는 기능 추가

-  ItemService 클래스 생성
	- findOne() : 상품 id로 repository에서 상품을 찾아다, responseDto로 응답
- ItemDetailResponse 수정 : 기존 더미 응답 대신 실제 Item 기반으로 생성하도록 함

ttasjwi committed:
[#53] Refactor: 메인이미지, 서브이미지 URL들을 가져오는 로직을 Item에 위임

- 기존 : ItemDetailResponse에서 아이템 이미지 목록을 필터링하는 로직을 가졌음
- 변경 후 : 상품의 메인이미지, 서브이미지 목록 가져오는 로직을 Item에 위임

ttasjwi committed:
[#32] Feat: 특정 섹션에 속한 상품 목록 가져오는 기능 추가

- findBySection : 상품 섹션명으로 repository에서 상품을 찾아다, responseDto로 응답
- ItemListElement : 기존엔 더미데이터로 응답했었으나, 실제로 Item 기반으로 생성하도록 변경
- ItemListResponse : ItemElementDto를 감싼 클래스

wheejuni commented on BE/src/main/java/codesquad/sidedish/domain/delivery/DeliveryRepository.java:
```diff
@@ -0,0 +1,61 @@
+
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("deliveryId", deliveryId);
+        return jdbcTemplate.query(selectDeliverySql, parameters, deliveryRowMapper()).stream().findFirst();
```
`.findFirst()` 한 후 `.orElseThrow()` 등 예외처리까지 여기서 마무리하면 어떨까 싶습니다.

ttasjwi replied:
이 부분은 서비스 로직 등에서 처리하도록 하는게 편할 것 같아서 우선 그대로 두었습니다.

wheejuni commented on BE/src/main/java/codesquad/sidedish/domain/item/Item.java:
```diff
@@ -0,0 +1,94 @@
+
+    private void validateReduceQuantity(int reduceNumber) {
+        if (reduceNumber > this.stock) {
+            throw new IllegalArgumentException(
```
별도의 예외가 있으면 좋겠습니다! 예외 이름은 뭐가 좋을까요...

street62 replied:
이 경우에는 `NotEnoughStockException`으로 정의했습니다!

wheejuni commented on BE/src/main/java/codesquad/sidedish/domain/item/Item.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    public int getMileage() {
+        return (int)(getDiscountedItemPrice() * 0.01);
```
`0.01` 이라는 숫자가 어떤 의미를 갖는지를 정확히 변수로써 나타내야 한다고 봅니다.

ttasjwi replied:
막 써놓고보니 매직넘버군요. 별도로 필드에 static 상수를 선언해봐야겠습니다.

wheejuni commented on BE/src/main/java/codesquad/sidedish/domain/item/Item.java:
```diff
@@ -0,0 +1,94 @@
+
+    public List<String> getMainImageUrls() {
+        return itemImages.stream()
+                .filter(itemImage -> itemImage.getItemImageType() == ItemImageType.MAIN)
```
`ItemImage` 클래스에 메인 이미지 여부를 판별하는 메소드가 있으면 좋겠군요.
게터를 불러서 꺼내온 다음 등호로 비교하지 말고요.

ttasjwi replied:
이미지에 있어 메인이미지냐, 서브이미지냐 판단하는 로직을 자체적으로 부여하면 자율성이 늘어날 것 같긴합니다. 이건 반영해보겠습니다!

wheejuni commented on BE/src/main/java/codesquad/sidedish/domain/item/Item.java:
```diff
@@ -0,0 +1,94 @@
+
+    public List<String> getSubImageUrls() {
+        return itemImages.stream()
+                .filter(itemImage -> itemImage.getItemImageType() == ItemImageType.SUB)
```
여기도 같은 의견 드립니다

street62 replied:
이 경우에는 `ItemNotFoundException`으로 정의했습니다!

wheejuni commented on BE/src/main/java/codesquad/sidedish/domain/member/Member.java:
```diff
@@ -36,5 +36,7 @@ public void initMemberId(Long memberId) {
     }
 
-
+    public void earnMileage(int mileage) {
+        this.mileage += mileage;
+    }
```
단순히 마일리지를 더하기 처리만 해주면 되는 걸까요? 이력 관리는 없어도 될까요?
회원의 마일리지 가감 이력을 모두 보관하고 있다가, 그 회원의 마일리지 합을 알고 싶으면 그걸 `.sum()` 하는 방향이 좋을 것 같은데..
어떻게 생각하세요?

ttasjwi replied:
이력관리... 이번 미션의 시간관계상 거기까지 바로 구현하기 힘들 수 있을 것 같습니다. 다른 것들을 먼저 구현해보고 시간적 여유가 있으면 반영해보도록 하겠습니다.  ㅜㅜ

wheejuni commented on BE/src/main/java/codesquad/sidedish/service/ItemService.java:
```diff
@@ -0,0 +1,37 @@
+
+    public ItemDetailResponse findOne(Long itemId) {
+        Item item = itemRepository.findById(itemId).orElseThrow(
+                () -> new NoSuchElementException("해당하는 id의 상품이 없습니다.")
```
여기도 이 상황에 맞는 예외가 있어야 할 것 같습니다!
프로젝트 전역으로 `404` 에 해당하는 예외를 설계하시는 것도 좋아요

street62 replied:
이 경우에는 ItemNotFoundException으로 정의했습니다!

wheejuni reviewed (changes requested):
수고하셨습니다 💯 
몇 가지 의견 남깁니다. 확인 및 반영 부탁드려요

ttasjwi committed:
[#54] Feat: 상품 조회 실패 예외 추가 및 대체

- 상품을 찾지 못 했을 때의 예외를 정의한 ItemNotFoundException을 정의했다.

ttasjwi committed:
[#54] Feat: 재고 부족 커스텀 예외 추가

- 재고량보다 구입하려는 수량이 더 많을 경우의 예외를 NotEnoughStockException로 별도로 정의했다.

ttasjwi committed:
[#54] Feat: 커스텀예외들에 대한 예외처리 로직 추가

- ErrorResult : 예외 응답을 위한 객체
- MyExceptionHandler : RestController 단에서 throw된 예외들을 처리하여 응답
- 예외 발생, 정상 처리 시 상태코드를 구분하여 응답하도록 함.

ttasjwi committed:
[#56] Refactor: ItemImage에 메인/서브이미지 여부를 판단하는 로직 추가

- 기존
    - ItemImage에서 getter를 통해 ItemImageType을 얻어와 타입을 비교했음.
    - ItemImage가 수동적
- 변경 후
    - ItemImage는 스스로가 메인이미지인지 서브이미지인지 응답할 수 있어야함. ItemImage에 isMainImage, isSubImage 메서드를 추가하여 메인이미지인지, 서브이미지인지 여부를 묻도록 함

ttasjwi committed:
[#57] Refactor: 마일리지 적립률 상수 추가

- 아이템 구입 시 마일리지 적립률 0.01을 명시적으로 static 상수로 선언
